### PR TITLE
Remove unencrypted task args from database

### LINKED ISSUE #4546: Remove unencrypted task arguments

### PR DESCRIPTION
**Pull request by mdellweg**
Do not merge before 3.40!

fixes #4546

### CODE REVIEW

File: pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    for task in Task.objects.filter(Q(args__is_null=False) | Q(kwargs__is_null=False)):
+        task.enc_args = task.args
+        task.enc_kwargs = task.kwargs
+        coreversion = task.versions.get("core")
+        if coreversion is None or parse_version(coreversion) < parse_version("3.34.0"):
+            task.versions["core"] = "3.34.0"
+
+        batch.append(task)
+        if len(batch) >= 500:
+            Task.bulk_update(batch, fields=["enc_args", "enc_kwargs", "versions"])
```
Comment (gerrod3) [inline pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py:22]: You forgot to clear the batch list here.

File: pulpcore/app/models/task.py [hunk 2]
```diff
@@ -83,12 +83,6 @@ class Task(BaseModel, AutoAddObjPermsMixin):
 
     error = models.JSONField(null=True)
 
-    # These fields should finally be removed in 3.40 by a migration that checks all components are
-    # running at least <version this landed> then copies their values to enc_(kw)args and adds
```
Comment (gerrod3) [inline pulpcore/app/models/task.py:87]: Did you forgot this check in the migration, or is it not needed?

**Review by gerrod3 — CHANGES_REQUESTED** [review]

File: pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py (hunk 1 above)
Comment (mdellweg) [reply pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py:22]: Right. Good find!

File: pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        Task.bulk_update(batch, fields=["enc_args", "enc_kwargs", "versions"])
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("core", "0113_headercontentguard"),
+    ]
+
+    operations = [
+        RequireVersion("core", "3.34.0"),
```
Comment (mdellweg) [inline pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py:34]: The check you were searching for is here.

File: pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+from django.db.models import Q
+from packaging.version import parse as parse_version
+
+from pulpcore.migrations import RequireVersion
+
+
+def encrypt_task_args_up(apps, schema_editor):
+    Task = apps.get_model("core", "Task")
+    batch = []
+    for task in Task.objects.filter(Q(args__is_null=False) | Q(kwargs__is_null=False)):
```
Comment (gerrod3) [inline pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py:13]: Do we need to use an `iterator` or `only` clause here? I assume some machines could have a large task database.

Comment (mdellweg) [reply pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py:13]: Are you saying, this queryset will be cached completely in memory once we access the first task? Will I ever fully understand this?

Comment (mdellweg) [reply pulpcore/app/migrations/0114_remove_task_args_remove_task_kwargs.py]: Reading https://docs.djangoproject.com/en/4.2/ref/models/querysets/#iterator , i now think that using `iterator()` will skip the caching altogether, which seems desirable here. I added it.

File: pulpcore/app/models/task.py (hunk 2 above)
Comment (mdellweg) [reply pulpcore/app/models/task.py:87]: See comment below.

File: pulpcore/tests/functional/api/using_plugin/test_filesystemexport.py [hunk 5]
```diff
@@ -15,8 +15,7 @@
 NUM_REPOS = 1
 NUM_EXPORTERS = 4
 
-if settings.DOMAIN_ENABLED:
-    pytest.skip("Domains do not support export.", allow_module_level=True)
+pytestmark = pytest.mark.skipif(settings.DOMAIN_ENABLED, reason="Domains do not support export.")
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/using_plugin/test_filesystemexport.py:18]: `pytest.skip(allow_module_level=True)` somehow started to fail with pytest-xdist...

**Review by ggainey — APPROVED** [review]

**Review by gerrod3 — APPROVED** [review]

**Pull request merged**